src/lock.rs

```diff
@@ -0,0 +1,174 @@
+// (c) Copyright 2023 Helsing GmbH. All rights reserved.
+
+use std::{collections::HashMap, sync::Arc};
+
+use eyre::{ensure, Context};
+use ring::digest;
+use semver::{Version, VersionReq};
+use serde::{Deserialize, Serialize};
+use tokio::fs;
+
+use crate::{
+    manifest::{Dependency, DependencyManifest},
+    package::{Package, PackageName},
+};
+
+pub const LOCKFILE: &str = "Proto.lock";
+
+/// Captures immutable metadata about a given package
+///
+/// It is used to ensure that future installations will use the exact same dependencies.
+#[derive(Serialize, Deserialize, Clone)]
+pub struct LockedPackage {
+    pub name: PackageName,
+    pub checksum: String,
```

Comment:
ideally we could put some stronger types around this? e.g. at this point, are we able to describe more about what hte checksum is? e.g. a Sha256Sum? Is the repository an HttpUrl?

Reply:
is the checksum specified as lower camel or upper camel?

Reply:
also - nit - checksum generally refers to values designed for purposes of detecting bit flip errors - more commonly in these lockfiles the value is called 'hash' or similar. Ideally if you know it's a sha256 hash, it's 'sha256_hash'.

Reply:
I actually tend to agree with all points. Unless @mara-schulke opposes I will change this tomorrow.

Reply:
I also agree

Reply:
Done. I actually used the term `digest` instead as that's a bit more formal / less ambiguous. Also changed how the hash is serialised so it encodes the hash algorithm, which helps with validation and may come in handy if we ever change the algorithm.